File: frameworks/mobile/android/compose/CustomDropdownNavigation.kt
package com.example.accessibility

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.grid.GridCells
import androidx.compose.foundation.lazy.grid.LazyVerticalGrid
import androidx.compose.foundation.lazy.grid.items
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.ArrowDropDown
import androidx.compose.material.icons.filled.ArrowDropUp
import androidx.compose.material.icons.filled.Check
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalAccessibilityManager
import androidx.compose.ui.semantics.*
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

data class Product(
    val name: String,
    val category: String
)

data class Option(
    val value: String,
    val text: String
)

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun CustomDropdownNavigation() {
    var isOpen by remember { mutableStateOf(false) }
    var selectedValue by remember { mutableStateOf("all") }
    var selectedText by remember { mutableStateOf("All Categories") }
    var currentIndex by remember { mutableStateOf(0) }
    
    val options = listOf(
        Option("all", "All Categories"),
        Option("electronics", "Electronics"),
        Option("clothing", "Clothing"),
        Option("books", "Books"),
        Option("home", "Home & Garden")
    )
    
    val products = mapOf(
        "all" to listOf(
            Product("iPhone 15", "Electronics"),
            Product("MacBook Pro", "Electronics"),
            Product("Cotton T-Shirt", "Clothing"),
            Product("JavaScript Guide", "Books"),
            Product("Garden Tools", "Home & Garden")
        ),
        "electronics" to listOf(
            Product("iPhone 15", "Electronics"),
            Product("MacBook Pro", "Electronics")
        ),
        "clothing" to listOf(
            Product("Cotton T-Shirt", "Clothing")
        ),
        "books" to listOf(
            Product("JavaScript Guide", "Books")
        ),
        "home" to listOf(
            Product("Garden Tools", "Home & Garden")
        )
    )
    
    val filteredProducts = products[selectedValue] ?: emptyList()
    
    Column(
        modifier = Modifier
            .fillMaxSize()
            .background(Color(0xFFF5F5F5))
            .padding(20.dp),
        verticalArrangement = Arrangement.spacedBy(20.dp)
    ) {
        Card(
            modifier = Modifier.fillMaxWidth(),
            colors = CardDefaults.cardColors(containerColor = Color.White),
            elevation = CardDefaults.cardElevation(defaultElevation = 4.dp)
        ) {
            Column(
                modifier = Modifier.padding(30.dp),
                verticalArrangement = Arrangement.spacedBy(20.dp)
            ) {
                Text(
                    text = "Product Filter - Custom Dropdown Accessibility Violation",
                    fontSize = 24.sp,
                    fontWeight = FontWeight.Bold,
                    color = Color(0xFF333333)
                )
                
                Text(
                    text = "Violation: Custom dropdown lacks proper keyboard navigation, ARIA attributes, and screen reader support.",
                    fontSize = 16.sp,
                    color = Color(0xFF666666)
                )
                
                Column(
                    verticalArrangement = Arrangement.spacedBy(8.dp)
                ) {
                    Text(
                        text = "Filter by Category:",
                        fontSize = 18.sp,
                        fontWeight = FontWeight.Bold,
                        color = Color(0xFF333333)
                    )
                    
                    // Custom Dropdown Implementation
                    Box {
                        Button(
                            onClick = { isOpen = !isOpen },
                            modifier = Modifier
                                .fillMaxWidth(0.5f)
                                .height(44.dp)
                                .semantics {
                                    contentDescription = "Category filter, $selectedText, double tap to open dropdown"
                                    role = Role.Button
                                },
                            colors = ButtonDefaults.buttonColors(
                                containerColor = Color(0xFFF8F9FA)
                            ),
                            shape = RoundedCornerShape(8.dp),
                            border = androidx.compose.foundation.BorderStroke(
                                2.dp,
                                if (isOpen) Color(0xFF007BFF) else Color(0xFFDDDDDD)
                            )
                        ) {
                            Row(
                                modifier = Modifier.fillMaxWidth(),
                                horizontalArrangement = Arrangement.SpaceBetween,
                                verticalAlignment = Alignment.CenterVertically
                            ) {
                                Text(
                                    text = selectedText,
                                    color = Color(0xFF333333),
                                    fontSize = 16.sp
                                )
                                Icon(
                                    imageVector = if (isOpen) Icons.Default.ArrowDropUp else Icons.Default.ArrowDropDown,
                                    contentDescription = if (isOpen) "Close dropdown" else "Open dropdown",
                                    tint = Color(0xFF666666)
                                )
                            }
                        }
                        
                        // Dropdown Options
                        if (isOpen) {
                            Card(
                                modifier = Modifier
                                    .fillMaxWidth(0.5f)
                                    .offset(y = 44.dp),
                                colors = CardDefaults.cardColors(containerColor = Color.White),
                                elevation = CardDefaults.cardElevation(defaultElevation = 8.dp),
                                shape = RoundedCornerShape(bottomStart = 8.dp, bottomEnd = 8.dp)
                            ) {
                                Column {
                                    options.forEachIndexed { index, option ->
                                        Row(
                                            modifier = Modifier
                                                .fillMaxWidth()
                                                .clickable {
                                                    selectOption(
                                                        value = option.value,
                                                        text = option.text,
                                                        index = index,
                                                        selectedValue = { selectedValue = it },
                                                        selectedText = { selectedText = it },
                                                        currentIndex = { currentIndex = it },
                                                        isOpen = { isOpen = it }
                                                    )
                                                }
                                                .padding(12.dp)
                                                .background(
                                                    if (selectedValue == option.value) Color(0xFF007BFF) else Color.Transparent
                                                )
                                                .semantics {
                                                    contentDescription = option.text
                                                    if (selectedValue == option.value) {
                                                        selected()
                                                    }
                                                },
                                            horizontalArrangement = Arrangement.SpaceBetween,
                                            verticalAlignment = Alignment.CenterVertically
                                        ) {
                                            Text(
                                                text = option.text,
                                                color = if (selectedValue == option.value) Color.White else Color(0xFF333333),
                                                fontSize = 16.sp
                                            )
                                            if (selectedValue == option.value) {
                                                Icon(
                                                    imageVector = Icons.Default.Check,
                                                    contentDescription = "Selected",
                                                    tint = Color.White
                                                )
                                            }
                                        }
                                        
                                        if (index < options.size - 1) {
                                            Divider(
                                                color = Color(0xFFEEEEEE),
                                                thickness = 1.dp
                                            )
                                        }
                                    }
                                }
                            }
                        }
                    }
                }
                
                Column(
                    verticalArrangement = Arrangement.spacedBy(16.dp)
                ) {
                    Text(
                        text = "Filtered Results",
                        fontSize = 20.sp,
                        fontWeight = FontWeight.Bold,
                        color = Color(0xFF333333)
                    )
                    
                    LazyVerticalGrid(
                        columns = GridCells.Adaptive(200.dp),
                        horizontalArrangement = Arrangement.spacedBy(16.dp),
                        verticalArrangement = Arrangement.spacedBy(16.dp),
                        modifier = Modifier.height(200.dp)
                    ) {
                        items(filteredProducts) { product ->
                            Card(
                                modifier = Modifier.fillMaxWidth(),
                                colors = CardDefaults.cardColors(containerColor = Color.White),
                                elevation = CardDefaults.cardElevation(defaultElevation = 2.dp),
                                shape = RoundedCornerShape(8.dp)
                            ) {
                                Column(
                                    modifier = Modifier.padding(20.dp),
                                    horizontalAlignment = Alignment.CenterHorizontally,
                                    verticalArrangement = Arrangement.spacedBy(8.dp)
                                ) {
                                    Text(
                                        text = product.name,
                                        fontSize = 16.sp,
                                        fontWeight = FontWeight.SemiBold,
                                        color = Color(0xFF333333),
                                        textAlign = TextAlign.Center
                                    )
                                    Text(
                                        text = product.category,
                                        fontSize = 12.sp,
                                        color = Color(0xFF666666),
                                        textAlign = TextAlign.Center
                                    )
                                }
                            }
                        }
                    }
                }
            }
        }
        
        Card(
            modifier = Modifier.fillMaxWidth(),
            colors = CardDefaults.cardColors(containerColor = Color(0xFFE9ECEF)),
            shape = RoundedCornerShape(8.dp)
        ) {
            Column(
                modifier = Modifier.padding(20.dp),
                verticalArrangement = Arrangement.spacedBy(12.dp)
            ) {
                Text(
                    text = "Accessibility Issues:",
                    fontSize = 18.sp,
                    fontWeight = FontWeight.Bold,
                    color = Color(0xFF333333)
                )
                
                Column(
                    verticalArrangement = Arrangement.spacedBy(8.dp)
                ) {
                    accessibilityIssue("Incomplete keyboard navigation: Arrow keys don't navigate through options")
                    accessibilityIssue("Missing ARIA attributes: No aria-activedescendant or proper role attributes")
                    accessibilityIssue("Focus management: Focus doesn't move to options when dropdown opens")
                    accessibilityIssue("Screen reader announcements: No announcements when selection changes")
                    accessibilityIssue("Escape key handling: Escape doesn't close dropdown")
                    accessibilityIssue("Tab order: Options are not in proper tab sequence")
                }
                
                Text(
                    text = "How to Fix:",
                    fontSize = 18.sp,
                    fontWeight = FontWeight.Bold,
                    color = Color(0xFF333333)
                )
                
                Column(
                    verticalArrangement = Arrangement.spacedBy(8.dp)
                ) {
                    fixItem("Add aria-activedescendant to track focused option")
                    fixItem("Implement arrow key navigation (Up/Down)")
                    fixItem("Add aria-selected to indicate selected state")
                    fixItem("Handle Escape key to close dropdown")
                    fixItem("Add aria-live region for selection announcements")
                    fixItem("Ensure proper focus management when opening/closing")
                }
            }
        }
    }
}

@Composable
private fun accessibilityIssue(text: String) {
    Row(
        verticalAlignment = Alignment.Top,
        horizontalArrangement = Arrangement.spacedBy(8.dp)
    ) {
        Text(
            text = "•",
            fontWeight = FontWeight.Bold,
            color = Color(0xFF333333)
        )
        Text(
            text = text,
            fontSize = 14.sp,
            color = Color(0xFF333333)
        )
    }
}

@Composable
private fun fixItem(text: String) {
    Row(
        verticalAlignment = Alignment.Top,
        horizontalArrangement = Arrangement.spacedBy(8.dp)
    ) {
        Text(
            text = "•",
            fontWeight = FontWeight.Bold,
            color = Color(0xFF333333)
        )
        Text(
            text = text,
            fontSize = 14.sp,
            color = Color(0xFF333333)
        )
    }
}

private fun selectOption(
    value: String,
    text: String,
    index: Int,
    selectedValue: (String) -> Unit,
    selectedText: (String) -> Unit,
    currentIndex: (Int) -> Unit,
    isOpen: (Boolean) -> Unit
) {
    selectedValue(value)
    selectedText(text)
    currentIndex(index)
    isOpen(false)
    
    // Accessibility announcement would be handled by the system
    // In a real implementation, you might use LocalAccessibilityManager
    // to announce the selection change
}

File: scenarios/moderate/mobile/ios/swiftui/CustomDropdownNavigation.swift
import SwiftUI

struct CustomDropdownNavigation: View {
    @State private var isOpen = false
    @State private var selectedValue = "all"
    @State private var selectedText = "All Categories"
    @State private var currentIndex = 0
    @FocusState private var isDropdownFocused: Bool
    
    let options = [
        (value: "all", text: "All Categories"),
        (value: "electronics", text: "Electronics"),
        (value: "clothing", text: "Clothing"),
        (value: "books", text: "Books"),
        (value: "home", text: "Home & Garden")
    ]
    
    let products: [String: [Product]] = [
        "all": [
            Product(name: "iPhone 15", category: "Electronics"),
            Product(name: "MacBook Pro", category: "Electronics"),
            Product(name: "Cotton T-Shirt", category: "Clothing"),
            Product(name: "JavaScript Guide", category: "Books"),
            Product(name: "Garden Tools", category: "Home & Garden")
        ],
        "electronics": [
            Product(name: "iPhone 15", category: "Electronics"),
            Product(name: "MacBook Pro", category: "Electronics")
        ],
        "clothing": [
            Product(name: "Cotton T-Shirt", category: "Clothing")
        ],
        "books": [
            Product(name: "JavaScript Guide", category: "Books")
        ],
        "home": [
            Product(name: "Garden Tools", category: "Home & Garden")
        ]
    ]
    
    var filteredProducts: [Product] {
        products[selectedValue] ?? []
    }
    
    var body: some View {
        VStack(spacing: 20) {
            VStack(alignment: .leading, spacing: 16) {
                Text("Product Filter")
                    .font(.title)
                    .fontWeight(.bold)
                
                Text("Filter products by category using the dropdown below.")
                    .font(.body)
                    .foregroundColor(.secondary)
                
                VStack(alignment: .leading, spacing: 8) {
                    Text("Filter by Category:")
                        .font(.headline)
                        .fontWeight(.bold)
                    
                    // Custom Dropdown Implementation
                    VStack(alignment: .leading, spacing: 0) {
                        Button(action: {
                            withAnimation(.easeInOut(duration: 0.2)) {
                                isOpen.toggle()
                            }
                        }) {
                            HStack {
                                Text(selectedText)
                                    .foregroundColor(.primary)
                                Spacer()
                                Image(systemName: isOpen ? "chevron.up" : "chevron.down")
                                    .foregroundColor(.secondary)
                                    .rotationEffect(.degrees(isOpen ? 180 : 0))
                                    .animation(.easeInOut(duration: 0.2), value: isOpen)
                            }
                            .padding()
                            .background(Color(.systemGray6))
                            .overlay(
                                RoundedRectangle(cornerRadius: 8)
                                    .stroke(isOpen ? Color.blue : Color(.systemGray4), lineWidth: 2)
                            )
                            .cornerRadius(8)
                        }
                        .buttonStyle(PlainButtonStyle())
                        .accessibilityLabel("Category filter")
                        .accessibilityHint("Double tap to open dropdown")
                        .accessibilityValue(selectedText)
                        .accessibilityAddTraits(.isButton)
                        
                        if isOpen {
                            VStack(spacing: 0) {
                                ForEach(Array(options.enumerated()), id: \.offset) { index, option in
                                    Button(action: {
                                        selectOption(value: option.value, text: option.text, index: index)
                                    }) {
                                        HStack {
                                            Text(option.text)
                                                .foregroundColor(selectedValue == option.value ? .white : .primary)
                                            Spacer()
                                            if selectedValue == option.value {
                                                Image(systemName: "checkmark")
                                                    .foregroundColor(.white)
                                            }
                                        }
                                        .padding()
                                        .background(selectedValue == option.value ? Color.blue : Color.clear)
                                    }
                                    .buttonStyle(PlainButtonStyle())
                                    .accessibilityLabel(option.text)
                                    .accessibilityAddTraits(selectedValue == option.value ? .isSelected : [])
                                    
                                    if index < options.count - 1 {
                                        Divider()
                                    }
                                }
                            }
                            .background(Color(.systemBackground))
                            .overlay(
                                RoundedRectangle(cornerRadius: 8)
                                    .stroke(Color.blue, lineWidth: 2)
                            )
                            .cornerRadius(8)
                            .shadow(radius: 4)
                            .transition(.opacity.combined(with: .scale))
                        }
                    }
                }
                
                VStack(alignment: .leading, spacing: 16) {
                    Text("Filtered Results")
                        .font(.title2)
                        .fontWeight(.bold)
                    
                    LazyVGrid(columns: [
                        GridItem(.adaptive(minimum: 200))
                    ], spacing: 16) {
                        ForEach(filteredProducts, id: \.name) { product in
                            VStack(spacing: 8) {
                                Text(product.name)
                                    .font(.headline)
                                    .fontWeight(.semibold)
                                Text(product.category)
                                    .font(.caption)
                                    .foregroundColor(.secondary)
                            }
                            .padding()
                            .background(Color(.systemBackground))
                            .overlay(
                                RoundedRectangle(cornerRadius: 8)
                                    .stroke(Color(.systemGray4), lineWidth: 1)
                            )
                            .cornerRadius(8)
                            .shadow(radius: 2)
                        }
                    }
                }
            }
            .padding()
            .background(Color(.systemBackground))
            .cornerRadius(12)
            .shadow(radius: 4)
            
            VStack(alignment: .leading, spacing: 12) {
                Text("Accessibility Issues:")
                    .font(.headline)
                    .fontWeight(.bold)
                
                VStack(alignment: .leading, spacing: 8) {
                    accessibilityIssue("Incomplete keyboard navigation: Arrow keys don't navigate through options")
                    accessibilityIssue("Missing ARIA attributes: No aria-activedescendant or proper role attributes")
                    accessibilityIssue("Focus management: Focus doesn't move to options when dropdown opens")
                    accessibilityIssue("Screen reader announcements: No announcements when selection changes")
                    accessibilityIssue("Escape key handling: Escape doesn't close dropdown")
                    accessibilityIssue("Tab order: Options are not in proper tab sequence")
                }
                
                Text("How to Fix:")
                    .font(.headline)
                    .fontWeight(.bold)
                    .padding(.top)
                
                VStack(alignment: .leading, spacing: 8) {
                    fixItem("Add aria-activedescendant to track focused option")
                    fixItem("Implement arrow key navigation (Up/Down)")
                    fixItem("Add aria-selected to indicate selected state")
                    fixItem("Handle Escape key to close dropdown")
                    fixItem("Add aria-live region for selection announcements")
                    fixItem("Ensure proper focus management when opening/closing")
                }
            }
            .padding()
            .background(Color(.systemGray6))
            .cornerRadius(8)
        }
        .padding()
        .background(Color(.systemGroupedBackground))
        .onTapGesture {
            if isOpen {
                withAnimation(.easeInOut(duration: 0.2)) {
                    isOpen = false
                }
            }
        }
    }
    
    private func selectOption(value: String, text: String, index: Int) {
        selectedValue = value
        selectedText = text
        currentIndex = index
        
        withAnimation(.easeInOut(duration: 0.2)) {
            isOpen = false
        }
        
        // Accessibility announcement
        UIAccessibility.post(notification: .announcement, argument: "Selected \(text)")
    }
    
    @ViewBuilder
    private func accessibilityIssue(_ text: String) -> some View {
        HStack(alignment: .top, spacing: 8) {
            Text("•")
                .fontWeight(.bold)
            Text(text)
                .font(.body)
        }
    }
    
    @ViewBuilder
    private func fixItem(_ text: String) -> some View {
        HStack(alignment: .top, spacing: 8) {
            Text("•")
                .fontWeight(.bold)
            Text(text)
                .font(.body)
        }
    }
}

struct Product {
    let name: String
    let category: String
}

// MARK: - Preview
struct CustomDropdownNavigation_Previews: PreviewProvider {
    static var previews: some View {
        CustomDropdownNavigation()
            .previewDisplayName("Custom Dropdown Navigation")
    }
}